frontend-react/src/components/Admin/AdminReceiverDashboard.tsx

```diff
@@ -0,0 +1,775 @@
+        this.countSuccess = this.countFailed = 0;
+    }
+
+    updateState(newResult: boolean) {
+        if (newResult) {
+            this.countSuccess++;
+            // true and we're in one of these states
+            this.currentState = [
+                SuccessRate.UNDEFINED,
+                SuccessRate.ALL_SUCCESSFUL,
+            ].includes(this.currentState)
+                ? SuccessRate.ALL_SUCCESSFUL
+                : SuccessRate.MIXED_SUCCESS;
+        } else {
+            this.countFailed++;
+            // false and we're in one of these states
+            this.currentState = [
+                SuccessRate.UNDEFINED,
+                SuccessRate.ALL_FAILURE,
+            ].includes(this.currentState)
+                ? SuccessRate.ALL_FAILURE
+                : SuccessRate.MIXED_SUCCESS;
+        }
+        return this.currentState;
+    }
+
+    reset() {
+        this.currentState = SuccessRate.UNDEFINED;
+        this.countSuccess = 0;
+        this.countFailed = 0;
+    }
+}
+
+const SUCCESS_RATE_CLASSNAME_MAP = {
+    [SuccessRate.UNDEFINED]: "success-undefined",
+    [SuccessRate.ALL_SUCCESSFUL]: "success-all",
+    [SuccessRate.ALL_FAILURE]: "failure-all",
+    [SuccessRate.MIXED_SUCCESS]: "success-mixed",
+};
+
+function dateAddHours(d: Date, h: number): Date {
+    const result = new Date(d); // copy value
+    result.setHours(result.getHours() + h);
+    return result;
+}
+
+// mostly for readably
+type DatePair = [Date, Date];
+
+function dateIsInRange(d: Date, range: DatePair): boolean {
+    return d >= range[0] && d < range[1];
+}
+
+/**
+ *  simple iterator to make other code more readable.
+ *  Usage:
+ *    for (let eachTimeSlot in (new TimeSlots([dateStart, dateEnd], 2)) { }
+ */
+interface IterateTimeSlots {
+    [Symbol.iterator]: () => Iterator<DatePair>;
+}
+
+class TimeSlots implements IterateTimeSlots {
+    private current: Date;
+    private readonly end: Date;
+    private readonly skipHours: number;
+
+    constructor(range: DatePair, skipHours: number = 2) {
+        this.current = range[0];
+        this.end = range[1];
+        this.skipHours = skipHours;
+    }
+
+    *[Symbol.iterator]() {
+        do {
+            yield [
+                this.current,
+                dateAddHours(this.current, this.skipHours),
+            ] as DatePair;
+            this.current = dateAddHours(this.current, this.skipHours);
+        } while (this.current < this.end);
+    }
+}
+
+/**
+ * build the dictionary with a special path+key
+ * @param dataIn
+ */
+function makeDictionary(dataIn: AdmConnStatusDataType[]): DataDictionary {
+    // empty case
+    if (Object.keys(dataIn).length === 0) {
+        return {};
+    }
+
+    const data: DataDictionary = {};
+    let needsSorting = false;
+
+    let lastkey = "";
+    for (let item of dataIn) {
+        const key = `${item.organizationName}|${item.receiverName}\t${item.connectionCheckStartedAt}`;
+        data[key] = item;
+        if (lastkey > key) {
```

Comment:
Yeah. This keys-determine-sort-order is a concept take over from [key-value databases](https://en.wikipedia.org/wiki/Key%E2%80%93value_database).

The `MainRender` does test that the data resorts correctly for `makeDictionary`. (The data in `receiverConnectionCheckResultId: 2396` is out of order in `DATA` for the test.)

But in the end, the code probably doesn't really need keys since keys:value pairs aren't actually sorted by keys in javascript like they are in other languages. (I actually forgot this detail when I planned the code. Too many languages crammed into my head.)

That said, sorting strings is crazy FAST. And the reason why ISO dates are laid out the way they are is specific to descending resolution Year > month > day > hour > min > sec WITH leading zeros so they sort correctly as strings.

Sort order for ISO string > ISO string is the same as ISO Date > ISO Date.

Reply:
That said, simplifying this is a pretty complex refactor. I'd like to save it for another round if that's ok.

I'd like to tackle speeding up filtering in a refactor as well.

But "perfect can be the enemy of better", so I'd like to get this checked in, then iterate on it.

I put it in #6235 to keep track of it.